Allow Hibernate Reactive to coexist with Agroal (and thus Flyway and Liquibase)

tomas1885 opened the pull request:
Don't create a blocking persistent unit when jdbc is explicitly disabled for hibernate datasource
fixes #33380
~~fixes issues #10716 #13425~~
This solution can be better, since it relied on excplicity setting jdbc.enabled=false

yrodiere reviewed (changes requested):
Thanks, I added a few comments below.

But I'd like to clarify this PR most likely does not fix #13425, as I suspect the problem is much more complex.

It could possibly fix #33380 and (less confident about that) #10716, but we need to have dedicated tests to be sure of that.

Also, and perhaps more importantly, I wonder whether you can't simply remove the check about JDBC datasources...

yrodiere commented:
> Also, and perhaps more importantly, I wonder whether you can't simply remove the check about JDBC datasources...

Yes, from what I can see this condition was added in https://github.com/quarkusio/quarkus/commit/81a059613f55b68689b828156ebe7a660b97e8a3, and in effect we just want to generate a Hibernate ORM PU when not using Hibernate Reactive, and when using Hibernate Reactive it would be fine to not generate one (ever).

tomas1885 commented on integration-tests/hibernate-reactive-panache-agroal/pom.xml:
```diff
@@ -0,0 +1,413 @@
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>quarkus-integration-test-hibernate-reactive-panache-agroal</artifactId>
```
Panache is really not necessary as this happens with just hibernate reactive, but I had to make sure because we use panache reactive in our project, this can be reduced to using hibernate reactive extension only

tomas1885 commented on integration-tests/hibernate-reactive-panache-agroal/src/main/java/io/quarkus/it/panache/reactive/agroal/Bug7721Entity.java:
```diff
@@ -0,0 +1,32 @@
+import jakarta.persistence.Entity;
+
+@Entity
+public class Bug7721Entity extends Bug7721EntitySuperClass {
```
Yeah, I was just doing a quick draft and wanted to get integration test running quickly, this isn't the final solution, just a quick way to get started

tomas1885 commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:
```diff
@@ -267,10 +253,39 @@ public void parsePersistenceXmlDescriptors(HibernateOrmConfig config,
+                    .stream()
+                    .map(it -> it.datasource.orElse(null))
+                    .filter(Objects::nonNull)
+                    .map(dataSource -> {
+                        var datasourceJdbcEnabledKeys = DataSourceUtil.dataSourcePropertyKeys(dataSource, "jdbc");
+                        //We have to iterate over each key separately since quarkus.datasource."name".jdbc and quarkus.datasource.name.jdbc would not return the same result
+                        return datasourceJdbcEnabledKeys.stream()
+                                .map(it -> ConfigUtils.getFirstOptionalValue(List.of(it), Boolean.class).orElse(true))
+                                .reduce(
+                                        (enabled, otherEnabled) -> enabled && otherEnabled)
+                                .orElse(true);
+                    }).reduce((dataSourceJdbcEnabled, anotherDataSourceJdbcEnabled) -> dataSourceJdbcEnabled
+                            && anotherDataSourceJdbcEnabled)
+                    .orElse(true);
+        }
         // If we have some blocking datasources defined, we can have an implied PU
-        if (jdbcDataSourcesBuildItem.size() == 0 && capabilities.isPresent(Capability.HIBERNATE_REACTIVE)) {
+        if ((jdbcDataSourcesBuildItem.size() == 0 || !jdbcDataSourceEnabled)
+                && capabilities.isPresent(Capability.HIBERNATE_REACTIVE)) {
```
I agree, I just tried to keep the "intrusion" to a minimum, as there might be some reason why this was an issue in the first place.

yrodiere commented:
> If we can assume that once hibernate reactive is preset disabling the default blocking PU is fine and doesn't have any other side effects

I'd expect so, yes. But CI will confirm this once the PR gets finalized.

> If we're already on the issue is it possible to also use hibernate and hibernate reactive?

I don't know. I doubt that, because that seems too easy and I'd expect more changes are necessary, but you're free to try.

Also, ideally I'd prefer that we clean up the code to have one "common" extension and two specific extensions (one for ORM, one for Reactive), because that would be a saner basis for future work. And also because it would improve usability:  for example it would make *much* more sense to have Hibernate Reactive properties start with `quarkus.hibernate-reactive.` rather than `quarkus.hibernate-orm.`, and would remove ambiguity in many places. All that kind of work is what we'll need before we can consider #13425 fixed. But unfortunately I don't have time for that right now...

toms-neema committed:
When Hibernate Reactive is enabled don't create an ImpliedBlockingPersistenceUnit even when a blocking datasource is configured.
Add integration test to validate co-existence of hibernate reactive and agroal.

tomas1885 commented:
@yrodiere  I'm trying to bring this back from PR Limbo, I went with your approach. But I got to say that I don't feel confident that any other unknown behavior didn't break (One that is not covered by tests). I believe a more preferable way would be exposing a build Item from the reactive datasource extension, similar to what the Agroal extension does, this would prevent searching for another extension properties, and would allow the make a more informed decision about which persistence units should be configured.

yrodiere commented on integration-tests/hibernate-reactive-mysql-agroal/pom.xml:
```diff
@@ -0,0 +1,305 @@
+        </profile>
+
+        <profile>
+            <id>docker-mariadb</id>
```
The name of the Maven project and the dependencies mention mysql, but this and the configuration below mention mariadb... there's something off?

tomas1885 replied:
@yrodiere This was actually taken from the hibernate-reactive-mysql integration test, so I assume there's some reasoning behind this. I just copied the integration test I deemed as minimal and added a blocking endpoint to make sure they can coexist.

yrodiere replied:
Alright I'll send a PR to fix that, then.

yrodiere replied:
#35837

yrodiere reviewed (commented):
>  I'm trying to bring this back from PR Limbo, I went with your approach

Thanks a lot! I just have one comment about the integration test module mixing MariaDB with MySQL for some reason... ?

> But I got to say that I don't feel confident that any other unknown behavior didn't break (One that is not covered by tests).

I can empathize with that, believe me. But we won't hold you responsible for someone else failing to provide proper test coverage for this extension. The HR extension is still in preview, so we'll proceed with your changes and if a regression happens, we'll add a test and fix it.

> I believe a more preferable way would be exposing a build Item from the reactive datasource extension, similar to what the Agroal extension does, this would prevent searching for another extension properties, and would allow the make a more informed decision about which persistence units should be configured.

I agree it's something to keep in mind for #13425

yrodiere committed:
Run hibernate-reactive-mysql-agroal on CI

yrodiere commented:
I just pushed one more commit to your branch so that the new IT module actually gets executed on CI.

tomas1885 commented:
@yrodiere Thanks, I wasn't aware of that, is it somewhere in the contrib doc?

yrodiere commented:
> @yrodiere Thanks, I wasn't aware of that, is it somewhere in the contrib doc?

Yes: https://github.com/quarkusio/quarkus/blob/main/CONTRIBUTING.md#frequently-asked-questions

> The native integration test for my extension didn't run in the CI
>
> In the interest of speeding up CI, the native build job native-tests have been split into multiple categories which are run in parallel. This means that each new extension needs to be configured explicitly in [native-tests.json](https://github.com/quarkusio/quarkus/blob/main/.github/native-tests.json) to have its integration tests run in native mode.

We should probably have made that more obvious though... like adding it to https://github.com/quarkusio/quarkus/blob/main/CONTRIBUTING.md#tests-and-documentation-are-not-optional . I'll send a PR.

I suppose the idea was that people are expected to inspect logs to check that their new ITs actually ran... unrealistic expectations if you ask me, but :shrug:

tomas1885 commented:
@yrodiere That's on me, totally missed it, I must admit that I didn't even ran the native test locally, I'll hope it passes. I'll take better care next time.

tomas1885 commented:
@yrodiere BTW, I tested Flyway + HR on this branch and it seems to work fine, I can add an IT for it if you think it's useful.

yrodiere commented:
The failure is yet another instance of the Metaspace problem we have with our test suite. Merging, thanks!
